packages/ckeditor5-utils/src/dom/position.ts

```diff
@@ -99,12 +100,44 @@ export function getOptimalPosition( { element, target, positions, limiter, fitIn
 
 	let bestPosition: Position;
 
+	// Get intersection of all scrollable ancestors of `target`.
+	const allScrollableAncestors = getScrollableAncestors( target as HTMLElement );
+	const ancestorsIntersection: Rect | null = allScrollableAncestors
+		.reverse()
+		.map( item => new Rect( item as HTMLElement ) )
+		.reduce( ( ancestorIntersectionAccumulator, ancestorRect ): Rect => {
+			if ( !ancestorIntersectionAccumulator || ancestorIntersectionAccumulator.top === undefined ) {
+				return ancestorRect;
+			}
+
+			if ( !ancestorRect || ancestorRect.top === undefined ) {
+				return ancestorIntersectionAccumulator;
+			}
+
+			return ancestorIntersectionAccumulator.getIntersection( ancestorRect ) as Rect;
+		} );
+
 	// @if CK_DEBUG_POSITION // RectDrawer.clear();
 	// @if CK_DEBUG_POSITION // RectDrawer.draw( targetRect, { outlineWidth: '5px' }, 'Target' );
 
 	const viewportRect = fitInViewport && getConstrainedViewportRect( viewportOffsetConfig ) || null;
 	const positionOptions = { targetRect, elementRect, positionedElementAncestor, viewportRect };
 
+	if ( ( !ancestorsIntersection ) || ( !targetRect.getVisible() ) ) {
+		// Default last position is `viewportHidden`.
+		return new PositionObject( positions.at( -1 )!, positionOptions );
```

Comment:
`at()` is supported in Safari since Mar 2022 https://caniuse.com/?search=Array.at. Knowing yearly release cycle of MacOS I'm wondering how many "old" installations of MacOS are still in use and how often would this throw. Let's stay on the safe side for a little longer, WDYT?